src/plurals.rs

```diff
@@ -5,62 +5,203 @@ pub enum Resolver<'a> {
     /// A function/closure manually supplied by the user.
     Function(Box<Fn(u64) -> usize>),
     /// A boolean expression
-    Expr(&'a str, Option<Ast<'a>>),
+    /// Use Ast::parse to get an Ast
+    Expr(Box<Ast<'a>>),
+}
+
+
+fn index_of<'b, 'c>(src: &'b str, pat: &'static str) -> Option<usize> {
+   src.chars().fold((None, 0, 0, 0), |(match_index, i, n_matches, paren_level), ch| {
+       if let Some(x) = match_index {
+           return (Some(x), i, n_matches, paren_level);
+       } else {
+           let new_par_lvl = match ch {
+               '(' => paren_level + 1,
+               ')' => paren_level - 1,
+                _ => paren_level
+           };
+
+           if Some(ch) == pat.chars().nth(n_matches) {
+               let length = n_matches + 1;
+               if length == pat.len() && new_par_lvl == 0 {
+                   (Some(i - n_matches), i + 1, length, new_par_lvl)
+               } else {
+                   (match_index, i + 1, length, new_par_lvl)
+               }
+           } else {
+               (match_index, i + 1, 0, new_par_lvl)
+           }
+       }
+    }).0
 }
 
 use self::Ast::*;
-#[derive(Debug)]
-enum Ast<'a> {
+#[derive(Debug, PartialEq)]
+pub enum Ast<'a> {
     /// A ternary expression
     /// x ? a : b
     ///
     /// the three Ast<'a> are respectively x, a and b.
-    Ternary(&'a Ast<'a>, &'a Ast<'a>, &'a Ast<'a>),
+    Ternary(Box<Ast<'a>>, Box<Ast<'a>>, Box<Ast<'a>>),
     /// The n variable.
     N,
     /// Integer literals.
     Integer(u64),
-    /// Boolean literals.
-    Bool(bool),
-    /// Comparison operators.
-    CompOp(&'a str, &'a Ast<'a>, &'a Ast<'a>),
-    /// && or || operators.
-    CombOp(&'a str, &'a Ast<'a>, &'a Ast<'a>),
+    /// Binary operators.
+    Op(&'a str, Box<Ast<'a>>, Box<Ast<'a>>),
     /// ! operator.
-    Not(&'a Ast<'a>),
+    Not(Box<Ast<'a>>),
 }
 
 impl<'a> Ast<'a> {
     fn resolve(&self, n: u64) -> usize {
         match *self {
-            Ternary(cond, ok, nok) => if cond.resolve(n) == 0 {
+            Ternary(ref cond, ref ok, ref nok) => if cond.resolve(n) == 0 {
                 nok.resolve(n)
             } else {
                 ok.resolve(n)
             },
             N => n as usize,
             Integer(x) => x as usize,
-            Bool(b) => b as usize,
-            CompOp(op, lhs, rhs) => (match op {
-                "==" => lhs.resolve(n) == rhs.resolve(n),
-                "!=" => lhs.resolve(n) != rhs.resolve(n),
-                ">=" => lhs.resolve(n) >= rhs.resolve(n),
-                "<=" => lhs.resolve(n) <= rhs.resolve(n),
-                ">" => lhs.resolve(n) > rhs.resolve(n),
-                "<" => lhs.resolve(n) < rhs.resolve(n),
+            Op(ref op, ref lhs, ref rhs) => match *op {
+                "==" => (lhs.resolve(n) == rhs.resolve(n)) as usize,
```

Comment:
Implementing an enum for `Op` would probably be a good idea and would also get rid of `unreachable!()`.